src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/EventPipeMetadataGenerator.cs

```diff
@@ -61,9 +64,48 @@ internal sealed class EventPipeMetadataGenerator
             long keywords,
             uint level,
             uint version,
+            EventOpcode opcode,
             EventParameterInfo[] parameters)
+
+        {
+            byte[]? metadataV1 = GenerateMetadata(eventId, eventName, keywords, level, version, parameters,
+                out bool hasUnsupportedParamTypes);
+            byte[]? metadataV2 = null;
+            if (hasUnsupportedParamTypes || opcode != EventOpcode.Info)
+            {
+                if (metadataV1 == null)
+                {
+                    // We bailed on metadata generation so we can't trust anything
+                    return null;
+                }
+
+                metadataV2 = GenerateMetadataV2(eventId, eventName, keywords, level, version, opcode, parameters);
+                if (metadataV2 != null)
+                {
+                    // TODO: I would like this to not allocate 2x what it needs to
```

Comment:
With your metadata append strategy I see little reason to stick with my original format scheme : ) What about a format like this:

````
                // metadata blob
                metadataLength   : 4 bytes
                eventID          : 4 bytes
                eventName        : (eventName.Length + 1) * 2 bytes
                keywords         : 8 bytes
                eventVersion     : 4 bytes
                level            : 4 bytes
                parameterCount   : 4 bytes
                v1 parameters    : parameterCount bytes
                one or more optional tags follows, keep reading optional tags until bytes read == metadataLength
                each optional tag is encoded as:
                payload_length   : 4 bytes - the number of bytes that encode the optional tag payload 
                kind             : 1 byte - an enumeration of different kind of optional tags. If a reader doesn't understand
                                      the kind it can use the length to skip ahead to the next tag
                tag_payload      : payload_length bytes 
               
                enum tag_kind
                {
                    tag_kind_opcode = 1
                    tag_kind_parameter = 2
                    // other tags can be added in the future
                }

                //payload for tag_kind = opcode
               opcode           : 4 bytes

                // payload for tag_kind = parameter
                parameter_blob  : variable length
````

For events with non-default opcode, the opcode tag adds 9 bytes. For events that had parameter types not encodable previously the overhead is 5 bytes per parameter. I'd guess nearly always this will be smaller than duplicating the metadata and it has some easy extensibility.

Reply:
I agree with the part about combining the common metadata and making the opcode optional, but the way you have it described it seems to imply that you would give partial v1 metadata for events with unsupported types, which wouldn't work as far as I can see.

What is your intention when we have an event like say FooEvent(int i, int[] intArray, char c)? On my first reading of your comment it seemed like you were suggesting that the v1 parameter count and parameter blob would contain args i and c, and then v2 would have the unsupported types. But this can't work because the order of the parameters matters.

Just to confirm, what you're actually proposing is sharing the metadata parts that can be shared, and then the metadata blobs themselves  (minus the length/id/etc) would be the same as they currently are?

So for FooEvent(int i, int[] intArray, char c), parameterCount would be 0 and v1 parameters would be empty?

Reply:
> So for FooEvent(int i, int[] intArray, char c), parameterCount would be 0 and v1 parameters would be empty?

Yeah that was my intent, sorry for the confusion. My expectation is that any given metadata blob would either use the v1 parameter schema, or it would use the optional parameter tags, but never both.

Reply:
Ok, with the changes I just pushed the format is changed to the way agreed on here. I also combined the two generatemetadata methods to avoid the double allocation